packages/stream-node/writer.js

```diff
@@ -0,0 +1,72 @@
+/* Adapts a Node.js stream to an Writer<Uint8Array>, where a writer stream is
+ * modeled as a hybrid async iterator + generator.
+ */
+
+// @ts-check
+/// <reference types="ses"/>
+
+const { details: X } = assert;
+
+/**
+ * Adapts a Node.js writable stream to a JavaScript
+ * async iterator of Uint8Array data chunks.
+ * Back pressure emerges from awaiting on the promise
+ * returned by `next` before calling `next` again.
+ *
+ * @param {import('stream').Writable} writer the destination Node.js writer
+ * @returns {import('@endo/stream').Writer<Uint8Array>}
+ */
+export const makeNodeWriter = writer => {
+  assert(
+    !writer.writableObjectMode,
+    X`Cannot convert Node.js object mode Writer to AsyncIterator<undefined, Uint8Array>`,
+  );
+
+  const finished = new Promise(resolve => {
+    writer.on('finish', () => {
+      resolve({ done: true, value: undefined });
+    });
+  });
+
+  const errored = new Promise((_, reject) => {
+    writer.on('error', reject);
+  });
+
+  const closed = new Promise(resolve => {
+    writer.on('close', () => {
+      resolve({ done: true, value: undefined });
+    });
+  });
+
+  /** @type {import('@endo/stream').Writer<Uint8Array>} */
+  const nodeWriter = harden({
+    /** @param {Uint8Array} value */
+    async next(value) {
+      return Promise.race([
+        new Promise(resolve => {
+          if (!writer.write(value)) {
+            writer.once('drain', resolve);
+          } else {
+            resolve(undefined);
+          }
+        }),
+        finished,
+        errored,
+        closed,
+      ]);
```

Comment:
I like the `race` approach. Wondering however about the order of these. In particular, should the `resolve` from `write` have precedence over the others?

Reply:
This could probably be addressed by having `finish`, `error`, and `close` race to resolve a single promise, but I’m less sure whether that would be correct.

Reply:
I meant to simply place `finished, errored, closed` before `new Promise` since `race` will respect order

Reply:
Also not sure of the relative order between those, in particular, it feels like `errored` should come first.

Reply:
I made the relative order moot by consolidating these three cases into a promise that gets resolved once by whichever event gets emitted first.